src/dialog-service.js

```diff
@@ -34,32 +34,54 @@ export class DialogService {
    * @param settings Dialog settings for this dialog instance.
    * @return Promise A promise that settles when the dialog is closed.
    */
-  open(settings?: Object): Promise<DialogResult> {
-    return this.openAndYieldController(settings)
-      .then((controller) => controller.result);
-  }
-
-  /**
-   * Opens a new dialog.
-   * @param settings Dialog settings for this dialog instance.
-   * @return Promise A promise that settles when the dialog is opened.
-   * Resolves to the controller of the dialog.
-   */
-  openAndYieldController(settings?: Object): Promise<DialogController> {
+  open(settings?: DialogSettings): Promise<OpenDialogResult | CloseDialogResult> {
     let childContainer = this.container.createChild();
     let dialogController;
-    let promise = new Promise((resolve, reject) => {
+    let closeResult: Promise<CloseDialogResult> = new Promise((resolve, reject) => {
       dialogController = new DialogController(childContainer.get(Renderer), _createSettings(settings), resolve, reject);
     });
     childContainer.registerInstance(DialogController, dialogController);
-    dialogController.result = promise;
-    dialogController.result.then(() => {
+
+    closeResult.then(() => {
       _removeController(this, dialogController);
     }, () => {
       _removeController(this, dialogController);
     });
-    return _openDialog(this, childContainer, dialogController)
-      .then(() => dialogController);
+
+    let openResult = _getViewModel(this.container, this.compositionEngine, childContainer, dialogController).then((instruction) => {
```

Comment:
I think there should be a task to somehow reduce the complexity in here a bit later down the road but looks good for now!